lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -84,6 +84,30 @@ static void getBackwardSliceSimple(Operation *rootOp,
   backwardSlice.remove(rootOp);
 }
 
+// Reimplemented and simplified from SliceAnalysis to use a worklist rather than
+// recursion and not consider nested regions. Also accepts a clone set
+// representing a backward dataflow slice. As soon as a use is found outside the
+// backward slice, return false. Otherwise, the whole forward slice is contained
+// in the backward slice, and return true.
+static bool getForwardSliceSimple(Operation *rootOp,
+                                  SetVector<Operation *> &opsToClone,
+                                  SetVector<Operation *> &forwardSlice) {
+  SmallVector<Operation *> worklist;
+  worklist.push_back(rootOp);
+
+  while (!worklist.empty()) {
+    Operation *op = worklist.pop_back_val();
+    forwardSlice.insert(op);
+    for (auto *user : op->getUsers()) {
+      if (!opsToClone.contains(user))
+        return false;
+      worklist.push_back(user);
+    }
+  }
+
+  return true;
+}
```

Comment:
Can we track a set of visited operations not to visit the same operations twice? It would reduce the worst case time complexity from O(N^2) to O(N). Also if there is a cyclic dependency, it might cause an infinite loop.

Reply:
@uenoku is [fb306c4](https://github.com/llvm/circt/pull/3522/commits/fb306c48ef77c0c5a25dcde2bb25d267815ed42d) what you had in mind? I think so, since I forgot to included the "visited" set from textbook BFS implementations. It also occurred to me we could try to share some cache between calls to try an memoize if certain instances share common forward dataflow sub-graphs, but I'm not sure how useful that would be in practice.

Reply:
That implementation makes sense to me but probably the following implementation is slightly more efficient. 
```suggestion
  worklist.push_back(rootOp);
  visited.insert(rootOp);

  while (!worklist.empty()) {
    Operation *op = worklist.pop_back_val();
    forwardSlice.insert(op);

    for (auto *user : op->getUsers()) {
      if (!opsToClone.contains(user))
        return false;
      if (visited.insert(op).second)
        worklist.push_back(user);
    }
  }
```

Reply:
Actually^ pebcak, I have fixed this.